File: templates/problems/13_dp_1d/12_partition_equal_subset/python/solution.py
def solve_can_partition(nums: list[int]) -> bool:
    """Check if the array can be partitioned into two equal-sum subsets.

    Time: O(n * sum), Space: O(sum)
    """
    total = sum(nums)
    if total % 2 != 0:
        return False
    target = total // 2
    dp = [False] * (target + 1)
    dp[0] = True
    for num in nums:
        for j in range(target, num - 1, -1):
            dp[j] = dp[j] or dp[j - num]
    return dp[target]

File: templates/problems/13_dp_1d/12_partition_equal_subset/typescript/solution.ts
/**
 * Check if the array can be partitioned into two equal-sum subsets.
 * Time: O(n * sum), Space: O(sum)
 */
export function solveCanPartition(nums: number[]): boolean {
    const total = nums.reduce((acc, n) => acc + n, 0);
    if (total % 2 !== 0) return false;
    const target = total / 2;
    const dp = new Array<boolean>(target + 1).fill(false);
    dp[0] = true;
    for (const num of nums) {
        for (let j = target; j >= num; j--) {
            dp[j] = dp[j] || dp[j - num];
        }
    }
    return dp[target];
}
